plugins/kafka/src/plugin.cpp

```diff
@@ -0,0 +1,410 @@
+        }
+      }
+    }
+    // Create the consumer.
+    auto client = consumer::make(*cfg);
+    if (!client) {
+      ctrl.diagnostics().emit(
+        diagnostic::error("failed to create consumer: {}", client.error())
+          .done());
+      return {};
+    };
+    if (auto value = cfg->get("bootstrap.servers")) {
+      VAST_INFO("kafka consumer connected to: {}", *value);
+    }
+    auto topic = args_.topic ? args_.topic->inner : default_topic;
+    VAST_INFO("kafka subscribes to topic {}", topic);
+    if (auto err = client->subscribe({topic})) {
+      ctrl.diagnostics().emit(
+        diagnostic::error("failed to subscribe to topic: {}", err).done());
+      return {};
+    }
+    // Setup the coroutine factory.
+    auto make
+      = [](loader_args args, consumer client) mutable -> generator<chunk_ptr> {
+      auto num_messages = size_t{0};
+      while (true) {
+        auto msg = client.consume(500ms);
+        if (!msg) {
+          co_yield {};
+          if (msg.error() == ec::timeout)
+            continue;
+          if (msg.error() == ec::end_of_input)
+            // FIXME: currently doesn't work for N partitions with N > 1.
+            // Upgrade to a counter and only break out of the loop once this
+            // signal has been received N times.
+            break;
+          VAST_ERROR(msg.error());
+          break;
+        }
+        auto payload = chunk::copy(msg->payload());
+        co_yield payload;
+        if (args.count && args.count->inner == ++num_messages)
+          break;
+      }
+    };
+    return make(args_, std::move(*client));
+  }
+
+  auto to_string() const -> std::string override {
+    auto result = name();
+    result += fmt::format(" --topic {}", args_.topic);
+    if (args_.count)
+      result += fmt::format(" --count {}", args_.count->inner);
+    if (args_.exit)
+      result += "--exit";
+    if (args_.offset)
+      result += fmt::format(" --offset {}", args_.offset->inner);
+    if (args_.options)
+      result += fmt::format(" --set {}", args_.options->inner);
+    return result;
+  }
+
+  auto name() const -> std::string override {
+    return "kafka";
+  }
+
+  auto default_parser() const -> std::string override {
+    return "json";
+  }
+
+  friend auto inspect(auto& f, kafka_loader& x) -> bool {
+    return f.object(x)
+      .pretty_name("kafka_loader")
+      .fields(f.field("args", x.args_), f.field("config", x.config_));
+  }
+
+private:
+  loader_args args_;
+  record config_;
+};
+
+struct saver_args {
+  std::optional<located<std::string>> topic;
+  std::optional<located<std::string>> key;
+  std::optional<located<std::string>> timestamp;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, saver_args& x) -> bool {
+    return f.object(x)
+      .pretty_name("saver_args")
+      .fields(f.field("topic", x.topic), f.field("key", x.key),
+              f.field("timestamp", x.timestamp));
+  }
+};
+
+class kafka_saver final : public plugin_saver {
```

Comment:
For the saver, you likely want batches of size 1; I'll write a follow-up story for this.